### PR TITLE
feat(plugin): ai prompt template

### PR DESCRIPTION
**Pull request by ttyS0e**
### Summary

This plugin adds the ability to provide tuned AI prompts to users, who only need to fill in the blanks with variable placeholders (in moustache format: `{{variable}}` )

**It is to be used in conjunction with `ai-proxy` plugin, as with other AI Family plugins.**

[AI-Proxy plugin](https://github.com/Kong/kong/pull/12207)

This allows prompt tuning to be started, or performed, by the experience LLM users in an organisation, and then consumed by anyone.

When activated, it looks for template references in the following forms:

**Chat:**
```json
{
	"messages": "{template://developer-chat}",
	"properties": {
		"language": "python",
		"program": "flask web server"
	}
}

```

**Prompt:**
```json
{
	"prompt": "{template://developer-prompt}",
	"properties": {
		"language": "python",
		"program": "flask web server"
	}
}
```

**Based on this gateway configuration:**
```yaml
_format_version: "3.0"

services:
- name: default
  host: localhost
  path: "/"
  port: 9900
  protocol: https
  routes:
  - name: openai-chat
    paths:
    - "~/openai/chat$"
    methods:
    - POST
    plugins:
    - name: ai-proxy
      config:
        route_type: "llm/v1/chat"
        auth:
          header_name: "Authorization"
          header_value: "Bearer REDACT"
        logging:
          log_statistics: true
          log_payloads: false
        model:
          provider: "openai"
          name: "gpt-3.5-turbo"
          options:
            max_tokens: 512
            temperature: 1.0
    plugins:
    - name: ai-prompt-template
      config:
        allow_untemplated_requests: true
        templates:
        - name: "developer-chat"
          template:  |-
            {
              "messages": [
                {
                  "role": "system",
                  "content": "You are a {{program}} expert, in {{language}} programming language."
                },
                {
                  "role": "user",
                  "content": "Write me a {{program}} program."
                }
              ]
            }
        - name: "developer-prompt"
          template:  |-
            {
              "prompt": "You are a {{language}} programming language expert. Write me a {{program}} program."
            }
```

It has been designed to work with other formats as soon as they are supported.

It also sanitises string inputs to ensure that json control characters are escaped, preventing arbitrary prompt injection.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - docs PR for all four "AI Family" pluigins is still being worked on separately.

### Issue reference

Internal project.

### CODE REVIEW

File: spec/03-plugins/41-ai-prompt-template/02-integration_spec.lua [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+          method = "POST",
+        })
+        
+        local body = assert.res_status(400, r)
+        local json = cjson.decode(body)
+
+        assert.same(json, { error = true, message = "could not find template name [developer-doesnt-exist]" } )
+      end)
+
+      it("errors with missing template parameter", function()
```
Comment (brentos) [inline spec/03-plugins/41-ai-prompt-template/02-integration_spec.lua:227]: Add a test for multiple missing parameters

File: kong/plugins/ai-prompt-template/templater.lua [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    sanitized_properties[k], _, err = sanitize_parameter(v)
+    if err then return nil, err end
+  end
+
+  local result = template.template:gsub(GSUB_REPLACE_PATTERN, sanitized_properties)
+  
+  -- find any missing variables
+  local errors = {}
+  local error_string
+  for w in (result .. ";"):gmatch(GSUB_REPLACE_PATTERN) do
```
Comment (brentos) [inline kong/plugins/ai-prompt-template/templater.lua:75]: Why concatenate the semicolon?

File: kong/plugins/ai-prompt-template/templater.lua [hunk 3]
```diff
@@ -0,0 +1,92 @@
+local _S = {}
+
+-- imports
+local fmt           = string.format
+--
+
+-- globals
+local GSUB_REPLACE_PATTERN = "{{([%w_]+)}}"
```
[the remark this answers is not in the document]
Comment (ttyS0e) [reply kong/plugins/ai-prompt-template/templater.lua:8]: Changed to underscore, typo.

File: kong/plugins/ai-prompt-template/templater.lua (hunk 2 above)
Comment (ttyS0e) [reply kong/plugins/ai-prompt-template/templater.lua:75]: I have... no idea why I did this. I've removed it.

**Comment by ttyS0e** [comment]
All addressed in https://github.com/Kong/kong/pull/12340

**Pull request closed without merging**